source/common/router/router.cc

```diff
@@ -124,7 +124,15 @@ uint64_t FilterUtility::percentageOfTimeout(const std::chrono::milliseconds resp
   return static_cast<uint64_t>(response_time.count() * TimeoutPrecisionFactor / timeout.count());
 }
 
-void FilterUtility::setUpstreamScheme(Http::RequestHeaderMap& headers, bool downstream_secure) {
+void FilterUtility::setUpstreamScheme(Http::RequestHeaderMap& headers,
+                                      absl::optional<bool> upstream_secure,
```

Comment:
make the upstream_secure be third parameter may be better and please add another bool flag to tell should we use the upstream_secure rather than using optional wrapper.

```

void FilterUtility::setUpstreamScheme(Http::RequestHeaderMap& headers, bool downstream_secure,
                                      bool upstream_secure, bool use_upstream_scheme);
```